### ISSUE TITLE
Edge NATS connection times out when Tailscale is enabled

### ISSUE THREAD

**Issue by leigh-johnson** [bitsy-ai/printnanny-os#196]
## **Describe the bug**
The Edge NATS connection is flaky AF when Tailscale service is enabled, leading to NATS timeout errors in the Mission Control dash.

I think what's happening is something like...

1. Tailscale attempts to resolve address for `foo.local`
2. Tailscale magic DNS resolution fails, since that's an mDNS address
3. Tailscale decides this must be an outbound Internet connection, forwards traffic to gateway device.
4. The gateway device can resolve the mDNS address, but NATs connection prooooobably timed out by now.

## **To Reproduce**
What are the steps to reproduce the issue?
1. Enable Tailscale service
6. Connect to your Tailnet
7. Refresh the Mission Control dashboard
8. Notice: at least one NATS request times out,

## **Expected behavior**

When Tailscale is enabled and you are connected to your Tailnet, use Tailscale's "magic DNS" instead of mDNS.

## References

Crash reports:

```
a1686bd5-fe92-4751-b3f0-e1b485686b9a
b9afcfd4-db51-4666-bfa6-25c313934977
```
```
-----------------------
Build Info:  |
-----------------------
IMAGE_NAME = printnanny-release-image-raspberrypi4-64-20230106203240
DISTRO_PRETTY = PrintNanny Linux 0.5.1 (Emerald)
VARIANT_NAME = PrintNanny OS Core
VARIANT_ID = core
DISTRO = printnanny
DISTRO_NAME = PrintNanny Linux
DISTRO_VERSION = 0.5.1
DISTRO_CODENAME = Emerald
ID_LIKE = BitsyLinux
HOME_URL = https://printnanny.ai
BUG_REPORT_URL = https://github.com/bitsy-ai/printnanny-os/issues
YOCTO_VERSION = 4.0.2
YOCTO_CODENAME = Langdale
MACHINE = raspberrypi4-64
TUNE_PKGARCH = cortexa72
-----------------------
Layer Revisions:      |
-----------------------
meta              = langdale:a3e3b740e140d036122f7b11e2ac452bda548444
meta-poky         = langdale:a3e3b740e140d036122f7b11e2ac452bda548444
meta-yocto-bsp    = langdale:a3e3b740e140d036122f7b11e2ac452bda548444
meta-raspberrypi  = libcamera-v0.0.3:104a8b9290ab4da92c8f693e2fdc42a6396975d2
meta-oe           = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-python       = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-multimedia   = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-bitsy        = main:957dfaa51204ce7739a3d222e50c5c33d656481e
meta-neural-network = master:fed68f0a659c9b6431985c156141dc28dc4c2275
meta-printnanny   = main:957dfaa51204ce7739a3d222e50c5c33d656481e
meta-networking   = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-filesystems  = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-initramfs    = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-webserver    = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-swupdate     = master:5e611f44db1b9333342ed61d77a3c3ed7be69c2b
workspace         = langdale:a3e3b740e140d036122f7b11e2ac452bda548444
meta-microcontroller = master:9f396e595b2d639be7f2b494f4a61e48d9495058
```

**Comment by leigh-johnson** [comment]
The good news is, this fix should be easy - add a setting to prefer using a shortname when rendering URLs. 
https://tailscale.com/kb/1081/magicdns/

Already put that on the roadmap, bumping priority since this blocks reasonable Tailscale usage: https://github.com/bitsy-ai/printnanny-os/issues/123

**Comment by leigh-johnson** [comment]
:rocket: Deploying a changeset that removes `Pi.fqdn` and composes URLs based on `Pi.network_settings.preferred_dns` field instead. Available in printnanny-api-client libraries >= 0.121.x

revs:
https://github.com/bitsy-ai/printnanny-webapp/pull/1079
https://github.com/bitsy-ai/printnanny-webapp/pull/1078

**Comment by leigh-johnson** [comment]
Woo, I'm pretty happy about where we're at with Tailscale support. Here's the current expected behavior:

- By default, you Raspberry Pi will be discoverable via mDNS and broadcast pn-v0-5.local
- If you are primarily using Tailscale, you can choose to prefer Tailscale's DNS. This will apply globally to all Raspberry Pis. https://printnanny.ai/settings/network/
- Enabling Tailscale does not interfere with mDNS resolution or avahi broadcast, so you can run split traffic:
```
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Joining mDNS multicast group on interface wlan0.IPv6 with address fe80::dea6:32ff:feb0:65d.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: New relevant interface wlan0.IPv6 for mDNS.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Joining mDNS multicast group on interface wlan0.IPv4 with address 192.168.1.78.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: New relevant interface wlan0.IPv4 for mDNS.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Joining mDNS multicast group on interface eth0.IPv6 with address fe80::dea6:32ff:feb0:65c.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: New relevant interface eth0.IPv6 for mDNS.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Joining mDNS multicast group on interface eth0.IPv4 with address 192.168.1.80.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: New relevant interface eth0.IPv4 for mDNS.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Joining mDNS multicast group on interface lo.IPv6 with address ::1.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: New relevant interface lo.IPv6 for mDNS.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Joining mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: New relevant interface lo.IPv4 for mDNS.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Network interface enumeration completed.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for fe80::dea6:32ff:feb0:65d on wlan0.*.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for fe80::d215:60cb:22fe:5cf4 on wlan0.*.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for 192.168.1.78 on wlan0.IPv4.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for 192.168.1.76 on wlan0.IPv4.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for fe80::dea6:32ff:feb0:65c on eth0.*.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for fe80::7ebd:1931:24:7e32 on eth0.*.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for 192.168.1.80 on eth0.IPv4.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for 192.168.1.69 on eth0.IPv4.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for ::1 on lo.*.
Jan 11 11:37:55 pn-v0-5 avahi-daemon[1266]: Registering new address record for 127.0.0.1 on lo.IPv4.
Jan 11 11:37:56 pn-v0-5 avahi-daemon[1266]: Server startup complete. Host name is pn-v0-5.local. Local service cookie is 3251755116.
```
```
pn-v0-5:~$ ifconfig -a
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.69  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::dea6:32ff:feb0:65c  prefixlen 64  scopeid 0x20<link>
        inet6 fe80::7ebd:1931:24:7e32  prefixlen 64  scopeid 0x20<link>
        ether dc:a6:32:b0:06:5c  txqueuelen 1000  (Ethernet)
        RX packets 6032  bytes 916870 (895.3 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 5047  bytes 459735 (448.9 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 88401  bytes 100691418 (96.0 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 88401  bytes 100691418 (96.0 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

tailscale0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1280
        inet 100.89.210.12  netmask 255.255.255.255  destination 100.89.210.12
        inet6 fd7a:115c:a1e0:ab12:4843:cd96:6259:d20c  prefixlen 128  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 631  bytes 62753 (61.2 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 785  bytes 67842 (66.2 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlan0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.76  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::dea6:32ff:feb0:65d  prefixlen 64  scopeid 0x20<link>
        inet6 fe80::d215:60cb:22fe:5cf4  prefixlen 64  scopeid 0x20<link>
        ether dc:a6:32:b0:06:5d  txqueuelen 1000  (Ethernet)
        RX packets 806  bytes 147605 (144.1 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 256  bytes 58721 (57.3 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Comment by leigh-johnson** [comment]
I'm going to close this issue for now, but will continue to kick the tires on the WebRTC stack while connected to a Tailnet.

The tl;dr of what I expect to happen is:

* If you're connected to the same LAN as your Raspberry Pi, ICE candidates should present LAN IPs (even when Tailscale tunnel is active)
* If you're connected to your Tailnet on the go, ICE candidates should present Tailnet IPs (since this will be the only route available). This should be a direct peer route **without** STUN relay, so double-check that we're not bouncing.

The second party (this should be a direct peer route **without** STUN relay) might not always be true if you're double NAT'd or your tailnet topology contains multiple subnets, so we'll have to do some more testing there. But so excited to see this progressing!
